src/Data/Singletons/Prelude/Show.hs

```diff
@@ -0,0 +1,229 @@
+
+  (:<>), (%:<>),
+  (:<>$), (:<>$$), (:<>$$$),
+
+  Shows, sShows,
+  ShowsSym0, ShowsSym1, ShowsSym2,
+
+  ShowListWith, sShowListWith,
+  ShowListWithSym0, ShowListWithSym1, ShowListWithSym2, ShowListWithSym3,
+
+  ShowChar, sShowChar,
+  ShowCharSym0, ShowCharSym1, ShowCharSym2,
+
+  ShowString, sShowString,
+  ShowStringSym0, ShowStringSym1, ShowStringSym2,
+
+  ShowParen, sShowParen,
+  ShowParenSym0, ShowParenSym1, ShowParenSym2,
+
+  ShowSpace, sShowSpace,
+  ShowSpaceSym0, ShowSpaceSym1,
+
+  ShowCommaSpace, sShowCommaSpace,
+  ShowCommaSpaceSym0, ShowCommaSpaceSym1,
+
+  AppPrec, sAppPrec, AppPrecSym0,
+
+  AppPrec1, sAppPrec1, AppPrec1Sym0
+  ) where
+
+import           Data.Kind (Type)
+import           Data.List.NonEmpty (NonEmpty)
+import           Data.Monoid ((<>))
+import           Data.Singletons.Prelude.Base
+import           Data.Singletons.Prelude.Instances
+import           Data.Singletons.Prelude.List
+import           Data.Singletons.Prelude.Ord
+import           Data.Singletons.Single
+import           Data.Singletons.TH
+import           Data.Singletons.TypeLits
+import qualified Data.Text as T
+
+import           GHC.TypeLits
+
+import qualified Prelude as P
+import           Prelude hiding (Show(..))
+
+import           Unsafe.Coerce (unsafeCoerce)
+
+-- | The promoted analogue of '(<>)' for 'Symbol's. This uses the special
+-- 'AppendSymbol' type family from "GHC.TypeLits".
+type a :<> b = AppendSymbol a b
+infixr 6 :<>
+
+-- | The singleton analogue of '(<>)' for 'Symbol's.
+(%:<>) :: Sing a -> Sing b -> Sing (a :<> b)
+sa %:<> sb =
+    let a  = fromSing sa
+        b  = fromSing sb
+        ex = someSymbolVal $ T.unpack $ a <> b
+    in case ex of
+         SomeSymbol (_ :: Proxy ab) -> unsafeCoerce (SSym :: Sing ab)
+infixr 6 %:<>
+
+type (:<>$$$) (x :: Symbol) (y :: Symbol) =
+    (:<>) x y
+instance SuppressUnusedWarnings (:<>$$) where
+  suppressUnusedWarnings = snd ((:<>$$###), ())
+data (:<>$$) (x :: Symbol) (y :: TyFun Symbol Symbol)
+  = forall arg. KindOf (Apply ((:<>$$) x) arg) ~ KindOf ((:<>$$$) x arg) => (:<>$$###)
+type instance Apply ((:<>$$) x) y = (:<>$$$) x y
+instance SuppressUnusedWarnings (:<>$) where
+  suppressUnusedWarnings = snd ((:<>$###), ())
+data (:<>$) (x :: TyFun Symbol (TyFun Symbol Symbol -> Type))
+  = forall arg. KindOf (Apply (:<>$) arg) ~ KindOf ((:<>$$) arg) => (:<>$###)
+type instance Apply (:<>$) x = (:<>$$) x
+
+-- | The @shows@ functions return a function that prepends the
+-- output 'Symbol' to an existing 'Symbol'.  This allows constant-time
+-- concatenation of results using function composition.
+type SymbolS = Symbol -> Symbol
+
+-- | GHC currently has no notion of type-level 'Char's, so we fake them with
+-- single-character 'Symbol's.
+type SChar = Symbol
+
+$(singletonsOnly [d|
+  class Show a where
+    showsPrec :: Nat -> a -> SymbolS
+    -- Intentionally renamed to avoid clashing with the Show class
+    show'     :: a -> Symbol
```

Comment:
In other places, `singletons` uses a trailing underscore to avoid naming conflicts. See `Data.Singletons.Prelude.Maybe.maybe_`. But remember to export the term-level name, then, too.